servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,420 @@
+      List<Long> children = Lists.newArrayList();
+      for (Lineage child : mLineageStore.getChildren(lineage)) {
+        children.add(child.getId());
+      }
+      info.children = children;
+      lineages.add(info);
+    }
+    return lineages;
+  }
+
+  /**
+   * It takes a checkpoint plan and waits for the lineage checkpointing service to checkpoint the
+   * lineages in the plan.
+   *
+   * @param plan the plan for checkpointing
+   */
+  public synchronized void waitForCheckpoint(CheckpointPlan plan) {
+    for (long lineageId : plan.getLineagesToCheckpoint()) {
+      Lineage lineage = mLineageStore.getLineage(lineageId);
+      // register the lineage file to checkpoint
+      for (LineageFile file : lineage.getOutputFiles()) {
+        // find the worker
+        long workerId = findStoringWorker(file);
+        if (workerId == -1) {
+          // the file is not on any worker
+          continue;
+        }
+        if (!mWorkerToCheckpointFile.containsKey(workerId)) {
+          mWorkerToCheckpointFile.put(workerId, Sets.<LineageFile>newHashSet());
+        }
+        mWorkerToCheckpointFile.get(workerId).add(file);
+      }
+    }
+  }
+
+  /**
+   * Polls the files to send to the given worker for checkpoint
+   *
+   * @param workerId the worker id
+   * @return the list of files.
+   * @throws FileDoesNotExistException
+   */
+  public synchronized List<CheckpointFile> pollToCheckpoint(long workerId)
+      throws FileDoesNotExistException {
+    List<CheckpointFile> files = Lists.newArrayList();
+    if (!mWorkerToCheckpointFile.containsKey(workerId)) {
+      return files;
+    }
+
+    List<Long> toRequestFilePersistence = Lists.newArrayList();
+    for (LineageFile file : mWorkerToCheckpointFile.get(workerId)) {
+      if (file.getState() == LineageFileState.COMPLETED) {
+        long fileId = file.getFileId();
+        toRequestFilePersistence.add(fileId);
+        List<Long> blockIds = Lists.newArrayList();
+        for (FileBlockInfo fileBlockInfo : mFileSystemMaster.getFileBlockInfoList(fileId)) {
+          blockIds.add(fileBlockInfo.blockInfo.blockId);
+        }
+
+        CheckpointFile toCheckpoint = new CheckpointFile(fileId, blockIds);
+        files.add(toCheckpoint);
+      }
+    }
+
+    requestFilePersistence(toRequestFilePersistence);
+    return files;
+  }
+
+  public synchronized void requestFilePersistence(List<Long> fileIds) {
+    if (!fileIds.isEmpty()) {
+      LOG.info("Request file persistency: " + fileIds);
+    }
+    for (long fileId : fileIds) {
+      mLineageStore.requestFilePersistence(fileId);
+    }
+    writeJournalEntry(new RequestFilePersistenceEntry(fileIds));
+    flushJournal();
+  }
+
+  private synchronized void requestFilePersistenceFromEntry(RequestFilePersistenceEntry entry) {
+    for (long fileId : entry.getFileIds()) {
+      mLineageStore.requestFilePersistence(fileId);
+    }
+  }
+
+  public synchronized void persistFiles(long workerId, List<Long> persistedFiles) {
```

Comment:
Yes, in a future PR we should push the async persist module down into the FS layer.